### PR TITLE
Do all validation steps before responding

### PR DESCRIPTION
**Pull request by jakobhellermann**
So that when you have multiple errors, you don't have to repeatedly change something and repost.

image

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -144,15 +144,17 @@ async def process_improvement_message(message: discord.Message, project: Optiona
...
             update_contributors(message.author, message.channel.id, project)
         else:
-            log.info(f"Warning {utils.detailed_user(message)} about {validation_result.log_text}")
+            log_messages = ", ".join(validation_result.log_text)
+            log.info(f"Warning {utils.detailed_user(message)} about {log_messages}")
             add_project_log(message)
             await message.add_reaction('❌')
             await message.add_reaction('⏭')
 
+            warnings="\n".join(validation_result.warning_text)
```
Comment (Kataiser) [inline main.py]: This should be moved a line up (whitespace after, not before)

File: validation.py [hunk 2]
```diff
@@ -12,13 +12,18 @@
 @dataclasses.dataclass
 class ValidationResult:
     valid_tas: bool
-    warning_text: Optional[str] = None
-    log_text: Optional[str] = None
+    warning_text: List[str]
+    log_text: List[str]
```
Comment (Kataiser) [inline validation.py]: You can just do list[str] now fyi

File: validation.py [hunk 3]
```diff
@@ -12,13 +12,18 @@
...
     finaltime: Optional[str] = None
     timesave: Optional[str] = None
     wip: bool = False
     sj_data: Optional[tuple] = None
 
+    def emit_error(self, warning: str, log_message: str):
```
Comment (Kataiser) [inline validation.py]: I don't really like this name, can't think of anything better but it's not really an error

File: tests.py
Comment (Kataiser) [inline tests.py:1]: I'll add a test for multiple validation failures

**Review by Kataiser — CHANGES_REQUESTED** [review]

**Comment by Kataiser** [comment]
Oh also can you make multiple failures in a response be in a markdown list please

File: validation.py (hunk 3 above)
Comment (jakobhellermann) [reply validation.py]: `emit_failed_check`?

File: main.py [hunk 4]
```diff
@@ -149,7 +149,7 @@ async def process_improvement_message(message: discord.Message, project: Optiona
             add_project_log(message)
             await message.add_reaction('❌')
             await message.add_reaction('⏭')
-            warnings="\n".join(validation_result.warning_text)
+            warnings = validation_result.warning_text[0] if len(validation_result.warning_text) == 1 else format_markdown_list(validation_result.warning_text)
```
Comment (jakobhellermann) [inline main.py]: is there a preferred way to break these long lines like
```py
warnings = validation_result.warning_text[0]
    if len(validation_result.warning_text) == 1
    else format_markdown_list(validation_result.warning_text)
```
or
```py
warnings = validation_result.warning_text[0] if len(validation_result.warning_text) == 1 else(
    format_markdown_list(validation_result.warning_text)
)
```

alternatively the `format_markdown_list` function could do the single-element check and this line would be `warnings = format_markdown_list(validation_result.warning_text)`.

**Comment by jakobhellermann** [comment]
image

File: validation.py (hunk 3 above)
Comment (Kataiser) [reply validation.py]: Sure

File: main.py (hunk 4 above)
Comment (Kataiser) [reply main.py]: I would say just break this into a full if-else instead of a ternary

Comment (jakobhellermann) [reply main.py]: Done

**Pull request merged**